[FEAT] oauth2 로그인 기능 추가

park-se-jun opened the pull request:
## PR Description 🗒️

OAuth2 기능 추가

- 테스트 아직 못함...


## 추가/변경 사항 및 설명 📝
- OAuth2 기능 관련 코드 추가 [참조한 레포](https://github.com/callicoder/spring-boot-react-oauth2-social-login-demo)
- spring security oauth2 설정
- sensitive.env 추가(prod 랑 같은 경로)
## To Reviewers 🙏

## Issue close ✂️
close #33

park-se-jun committed:
[CHORE] local용 db를 실행 시키는 docker compose 작성

park-se-jun committed:
[FEAT] flyway 적용 및 로컬 환경에서의 db 연결

park-se-jun committed:
[CHORE] application.yml 환경별 분리
- test 환경의 경우 h2사용하도록 분리
- local 환경의 경우 localhost:3306을 사용하도록 분리
- prod 환경일 경우 아직 작성 못함

park-se-jun committed:
[CHORE] prod 관련 applicatoin.yml 작성 및 cd 수정
- application-prod.yml 작성
- deploy.sh 에 profile 관련 옵션 추가
- deploy workflow에 prod의 db 관련 정보 설정

park-se-jun committed:
Merge pull request #57 from prgrms-web-devcourse/feature_#41

[FEAT] 플라이웨이 적용, test db docker compose 작성, 환경별 application.yml 분리

park-se-jun committed:
[FEAT] cors config 작성

- 아직 동작 x

park-se-jun committed:
[FEAT] basic security config 작성

- 아직 동작 x

park-se-jun committed:
Merge branch 'develop' into sejun

park-se-jun committed:
[CHORE] 임시로 pr과 deploy.yml 연결

park-se-jun committed:
[CHORE] pr 과 deploy.yml 연결 제거, 테스트용 db pr-check.yml에 추가

park-se-jun committed:
[CHORE] test시 test profile을 이용하도록 build 설정
- docker compose 관련 워크플로 삭제
- test 시 test profile 을 이용하도록 변경

park-se-jun committed:
[CHORE] pr에 deploy.yml 연결
- deploy.yml 트리거로 pr 연결
- pr-check 의 이름을 pr-check로 변경(원래 develop이었음)

park-se-jun committed:
[CHORE] shell script 수정 - profile을 prod로

park-se-jun committed:
[CHORE] 배포는 되지만 실행이 되지 않는 문제
- application-prod.yml 에 env를 file path로 import 하는데, file path는 커맨드 실행하는데를 기준으로 해서 배포시 잡지 못하는 문제가 있었음
- deploy.sh 에 -Dspring.config.import 옵션으로 덮어쓰기

park-se-jun committed:
[CHORE] spring.config.import 를 classpath로 변경한거에서 filepath로 되돌리기
- shell 파일 수정을 통해, 절대경로로 prod_db_info를 받게끔 했는데 application야믈 에는 classpath로 되어 잇었

park-se-jun committed:
[CHORE] deploy.yml에 임시로 pr 트리거 걸어 뒀던 것 제

park-se-jun committed:
[FEAT] 가장 기본적인 oauth2설정

park-se-jun committed:
[FIX] 오타 수정

park-se-jun committed:
[FEAT] 중간 때 참조 했던 코드들 붙여넣기( 이걸로 안할 수도 있음)

park-se-jun committed:
[CHORE] OAuth2 login 시 필요한 application.yml 수정 및 그에 해당하는 properties 생성
- client-id, client-secret 같이 필요한 정보들 은닉화 및 추가
- 그외 사용하는 정보 application에 추가
- 추가한 정보 properties로 생성

park-se-jun committed:
[DEL] 지원하지 않는  OAuth2 Userinfo 삭제

park-se-jun committed:
[DEL] Member의 refreshToken 관련 assertion 삭제
- refreshToken은 처음에 null 일 수 있음

park-se-jun committed:
[FEAT] TokenProvider를 우리의 코드에 맞게 수정

park-se-jun committed:
[FEAT] SecurityConfig수정
- h2-consle 관련 설정
- authorizationRequestRepository 설정
- SucessHandler 설정
- TokenFilter 설정

park-se-jun committed:
[FEAT] socialinfo assertion 잘못 되 있던거 수정

park-se-jun committed:
[FEAT] OAuth2 인증(Authentication) 성공시 DB 저장 로직 구현
- Authentication 성공시 MemberRepository에서 찾아봄
- 없다면 DB에 생성
- 있다면, 있는거 떠내서 씀
- 해당하는 userId로 AccessToken 만듦

park-se-jun committed:
[FEAT] 요청이 들어왔을 때 AccessToken을 검증하는 필터를 코드 상황에 맞게 수정

park-se-jun committed:
[FEAT] CORS 설정 적용되도록 변경
- @Configuration 어노테이셔 추가
- allowedOrigins를 @Value로 꽂아줌

park-se-jun committed:
[FEAT] 리프레스 토큰 관련 코드 추가

park-se-jun committed:
CHORE] 혹시 모를 front, back 주소 관련 application yml 수정

park-se-jun committed:
[CHORE] sensitive.env 추가에 따른 workflow 수정

park-se-jun committed:
sejun과 feature_#33의  conflict 해소 및 merge

park-se-jun committed:
Merge pull request #84 from prgrms-web-devcourse/feature_#33

[FEAT] OAuth2로 로그인 기능 추가 및 설정

park-se-jun committed:
[FIX] develop 과 sejun merge 해소

park-se-jun committed:
[FIX] code format 맞추기

twotwobread commented on src/main/java/com/example/demo/config/WebMVcConfig.java:
```diff
@@ -0,0 +1,25 @@
+	private String[] allowedOrigins;
+
+	@Override
+	public void addCorsMappings(CorsRegistry registry) {
```
CORS 설정 관련해서 얘기해보려고 했는데 좋습니다!

thguss replied:
인정합니다 따봉 👍

twotwobread commented on src/main/java/com/example/demo/security/oauth2/user/OAuth2UserInfoFactory.java:
```diff
@@ -0,0 +1,19 @@
+        if(registrationId.equalsIgnoreCase(Social.GOOGLE.toString())) {
+            return new GoogleOAuth2UserInfo(attributes);
+        } else {
+            throw new OAuth2AuthenticationException("Sorry! Login with " + registrationId + " is not supported yet.") {
```
이 예외는 어디서 잡히나요?

park-se-jun replied:
글로벌 컨트롤러 어드바이스에 추가 했습니다.

twotwobread commented on src/main/java/com/example/demo/security/oauth2/HttpCookieOAuth2AuthorizationRequestRepository.java:
```diff
@@ -0,0 +1,50 @@
+
+@Component
+public class HttpCookieOAuth2AuthorizationRequestRepository implements AuthorizationRequestRepository<OAuth2AuthorizationRequest> {
+    public static final String OAUTH2_AUTHORIZATION_REQUEST_COOKIE_NAME = "oauth2_auth_request";
```
public인 이유가 뭔가요?? 외부에서 사용하는 상수들인가요?

park-se-jun replied:
넵!

park-se-jun committed:
[CHORE] code format 맞추기

twotwobread commented on src/main/java/com/example/demo/security/oauth2/OAuth2AuthenticationSuccessHandler.java:
```diff
@@ -0,0 +1,112 @@
+		long memberId = joinMember((OAuth2AuthenticationToken)authentication);
+
+		String accessToken = tokenProvider.createAccessToken(memberId);
+		String refreshToken = tokenProvider.createRefrshToken(memberId);
```
궁금한 점은 인증에 성공하면 매번 refreshToken과 accessToken을 만드는 건가요?

thguss replied:
로그인은 할 때마다 access token과 refresh token 모두 업데이트 시켜주는 것으로 알고 있습니다!! 제가 이해하기로는 세준님의 로직도 이와 동일한 것 같은데 맞나요?-?

park-se-jun replied:
로직을 확인 해 보니 oauth2 인증이 성공할 때 마다(OAuth2Authentication 이 성공시) 새로 갱신해 주는 것 입니당.

p.s refrehToken을 이용한 토큰 재발급 로직이 빠져있는 상태긴 합니다. 다음 스프린트 중 시간이 남는다면, 진행하도록 하겠습니다

twotwobread commented on src/main/java/com/example/demo/service/MemberService.java:
```diff
@@ -0,0 +1,57 @@
+			element.setRefreshTken(refreshToken);
+			return element;
+		}).orElseThrow(
+			() -> new EntityNotFoundException("찾을 수 없는 유저 입니다")
```
ExceptionMessage를 이용하여 메시지를 넣어주면 좋을 것 같습니다.

thguss replied:
오 매의 눈 👀

park-se-jun replied:
반영 완료

twotwobread commented on src/main/java/com/example/demo/security/TokenProvider.java:
```diff
@@ -0,0 +1,83 @@
+	public boolean isExpiredToken(String authToken){
+
+		try {
+				Jwts.parser().setSigningKey(appProperties.getAuth().getTokenSecret()).parseClaimsJws(authToken);
```
들여쓰기 수정이 필요한 거 같습니다.

twotwobread commented on src/main/java/com/example/demo/security/oauth2/OAuth2AuthenticationSuccessHandler.java:
```diff
@@ -0,0 +1,112 @@
+                .map(Cookie::getValue);
+
+        if(redirectUri.isPresent() && !isAuthorizedRedirectUri(redirectUri.get())) {
+            throw new RuntimeException(
```
마땅한 Exception이 없다면 커스텀을 쓰는게 좋을 것 같습니다!

park-se-jun replied:
임시로 해둔 것 입니당.todo 로 해놓고 나중에 바꾸겠슴다~~

twotwobread reviewed (commented):
수고많았습니다 코멘트 남겼습니다!!

thguss reviewed (approved):
세상에,, 수고하셨습니다,, 👍
밖이라서 전체 꼼꼼하게는 못 보고, 훑어보고 수영님 코멘트를 중점으로 보고 리뷰 남겼습니다. 하하 ㅎㅁㅎ

park-se-jun committed:
[DEL] refreshToken 관련 검증 테스트 삭제

park-se-jun committed:
[FIX] 모든 요청에 대해 인증을 요구 하지 않도록 변경

park-se-jun committed:
[FIX] Genre 관련 controller 테스트가 Junit4 기반으로 되어 있던거 수정
- Junit4 기반으로 되어 있어서 테스트 인식을 하지 못하던 오류 수정

park-se-jun committed:
[FIX] Security 적용 과정에서 테스트가 실패하는 부분 수정
- TokenAuthenticationFilter를 스캔 대상에서 제외
- 위의 필터를 생성하기 위해선 TokenProvider와 AppProperties가 필요한데 이를 넣어주기 보단 필터 자체를 스캔 안하도록 변경

park-se-jun committed:
[FIX] 아직 테스트 하지 못한코드들을 jacoco의 exclude로 등록
- security 코드 전반
- security 코드에서 사용하는 util

park-se-jun committed:
[FEAT] OAuth2AuthenticationException 핸들링 포인트 추가

park-se-jun committed:
[FEAT] assignRefreshToken의 에러 처리시 ExceptionMessage로 처리

park-se-jun committed:
[CHORE] 후에 좀더 의미있는 예외보 바꾸기 위해 todo 추가

twotwobread reviewed (approved):
정말 수고많으셨습니다!!
